File: Python/student_attendance_record_1.py
# https://leetcode.com/problems/student-attendance-record-i/description/
from enum import StrEnum

from utils.examination import Examiner, Testcase

class Record(StrEnum):
    ABSENT = 'A'
    LATE = 'L'
    PRESENT = 'P'

class Solution:
    def checkRecord(self, s: str) -> bool:
        is_eligible: bool = True
        absent_times: int = 0
        late_times_window: int = 0
        for record in s:
            if Record.ABSENT == record:
                absent_times += 1
                if 1 < absent_times:
                    is_eligible = False
                    break
            if Record.LATE == record:
                late_times_window += 1
                if 2 < late_times_window:
                    is_eligible = False
                    break
            else:
                late_times_window = 0
        return is_eligible

def examine() -> None:
    Examiner(Solution(), 'checkRecord').run((
        Testcase(('PPALLP',), True),
        Testcase(('PPALLL',), False),
    ))

File: TypeScript/solutions/student-attendance-record-1.ts
// https://leetcode.com/problems/student-attendance-record-i/description/
import { Examiner, Testcase } from '../utils';


function checkRecord(s: string): boolean {
    let absentTimes = 0, lateWindow = 0;
    for (const record of s) {
        if (record === 'A') {
            absentTimes++;
            if (absentTimes > 1) return false;
        }
        if (record === 'L') {
            lateWindow++;
            if (lateWindow > 2) return false;
        } else {
            lateWindow = 0;
        }
    }
    return true;
}

export function examine(): void {
    new Examiner(checkRecord).run([
        new Testcase(['PPALLP'], true),
        new Testcase(['PPALLL'], false),
    ]);
}
